src/govuk/components/button/button.mjs

```diff
@@ -1,20 +1,46 @@
+import { mergeConfigs, normaliseDataset } from '../../common.mjs'
 import '../../vendor/polyfills/Event.mjs' // addEventListener and event.target normaliziation
 import '../../vendor/polyfills/Function/prototype/bind.mjs'
 
 var KEY_SPACE = 32
 var DEBOUNCE_TIMEOUT_IN_SECONDS = 1
 
-function Button ($module) {
+/**
+ * JavaScript enhancements for the Button component
+ *
+ * @class
+ * @param {HTMLElement} $module - The element this component controls
+ * @param {Object} config
+ * @param {Boolean} [config.preventDoubleClick=false] - Whether the button should prevent double clicks
+ */
+function Button ($module, config) {
+  if (!$module) {
+    return this
+  }
+
   this.$module = $module
   this.debounceFormSubmitTimer = null
+
+  var defaultConfig = {
+    preventDoubleClick: false
+  }
+  this.config = mergeConfigs(
+    defaultConfig,
+    config || {},
+    normaliseDataset($module.dataset)
+  )
 }
 
 /**
  * Initialise component
  */
 Button.prototype.init = function () {
+  if (!this.$module) {
```

Comment:
Asking for clarification: 

At the start of the module we're checking for the existence of `$module`, and only if it exists does the code to assign `$module` to `this.$module` run.

Here we are then checking for the existence of `this.$module`, however wouldn't the previous steps ensure that this statement would never evaluate to false?

Reply:
Unfortunately we need that extra check  in `init` because the first check only shortcuts the constructor, to safeguard the access ot the module's `dataset`. The `addEventListener` calls happen inside `init()` fail if the component was instantiated with an `undefined`/`null` element.

Reply:
If the constructor was called with an `undefined` element, wouldn't we want to immediately halt execution anyway?

Reply:
Sorry Ollie, I was commenting on an old version of the page and hadn't seen you'd replied.

> I don't think there's anything we can do from the constructor to stop the `init` function from being called – that's happening outside of the object (`new govukButton($module).init()`).

We could always make `init()` a no-op in the constructor. That would keep everything in the same place. That'd actually make the thing quite portable from component to component. 

```js
if (!$module) {
   // Make `init` a no-op to avoid potentially breaking code at initialisation
   // from lack of HTML element 
   this.init = function(){}
   return this
}
```

Reply:
I'll actually update the PR for that as:
- it'll avoir the awkward second check that prompted @querkmachine's question
- it'll be easy to pass from one component to another

I'll open a second PR to amend the components we've already made configurable.

Reply:
I'm not 100% sure about this from a code readability perspective – appreciate at the minute the init function is near the constructor but I think it'd be easy to miss and I don't think I'd expect a class method to be redefined in this way…

If I was writing this I think I'd stick with the two guard clauses we had before, especially if we plan to get rid of the `init` function in the short-to-mid term.

Reply:
Happy to reset to the previous commit, it's not a solution I was particularly attached to and it'll become irrelevant when we remove the `init` function. Just thought it'd make sense to keep everything in a single place instead of the double check, especially as it's for handling an "error" scenario.